### PR TITLE
Issue 18210

### PR DESCRIPTION
**Pull request by dsilvam**
The code in `NavTool.getNav`method was modifying the cache-stored copy of the SYSTEM_FOLDER, by replacing the hostId from SYSTEM_HOST to whatever the passed in for the method is. This was generating an issue in the validation when creating htmlpages. Now a defensive copy is made to avoid mutating the cache-stored copy.

### CODE REVIEW

File: dotCMS/src/main/java/com/dotcms/rendering/velocity/viewtools/navigation/NavTool.java [hunk 1]
```diff
@@ -77,13 +79,24 @@ protected NavResultHydrated getNav(Host host, String path, long languageId, User
             path = path.substring(0, path.lastIndexOf("/"));
         }
 
-        Folder folder = !path.equals("/") ? APILocator.getFolderAPI()
+        final Folder originalFolder = !path.equals("/") ? APILocator.getFolderAPI()
```
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotcms/rendering/velocity/viewtools/navigation/NavTool.java:82]: Codacy Issue found: [Position literals first in String comparisons](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5378894)

Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotcms/rendering/velocity/viewtools/navigation/NavTool.java:82]: Codacy Issue found: [Position literals first in String comparisons](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5378894)

File: dotCMS/src/integration-test/java/com/dotcms/rendering/velocity/viewtools/navigation/NavToolTest.java [hunk 2]
```diff
@@ -204,8 +205,15 @@ public void testRootLevelNavigation_WhenOneFileAssetIsShownOnMenu() throws Excep
                     .getNav(site, systemFolder.getPath(), 1, user);
             assertNotNull(navResult);
 
+            /* method below `findShowOnMenuUnderFolder` expects a mutated version of SYSTEM_FOLDER with the hostId altered.
+               So let's create a defensive copy of the SYSTEM_FOLDER, modify it and pass it to the method.
+            */
+            Folder modifiedSystemFolder = new Folder();
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/rendering/velocity/viewtools/navigation/NavToolTest.java:211]: Codacy Issue found: [Local variable 'modifiedSystemFolder' could be declared final](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5378894)

File: dotCMS/src/main/java/com/dotcms/rendering/velocity/viewtools/navigation/NavTool.java [hunk 3]
```diff
@@ -77,13 +79,24 @@ protected NavResultHydrated getNav(Host host, String path, long languageId, User
...
+
+        // make a defensive copy to avoid mutating cached version
+        Folder folder = new Folder();
+        try {
+            BeanUtils.copyProperties(folder, originalFolder);
+        } catch (IllegalAccessException | InvocationTargetException e) {
+            folder = originalFolder;
+            Logger.warnAndDebug(NavTool.class,"Defensive copy failed. Using original object", e);
+        }
+
```
Comment (freddyucv) [inline dotCMS/src/main/java/com/dotcms/rendering/velocity/viewtools/navigation/NavTool.java:99]: maybe it should go into a private method

File: dotCMS/src/integration-test/java/com/dotcms/rendering/velocity/viewtools/navigation/NavToolTest.java [hunk 4]
```diff
@@ -441,4 +449,36 @@ public void testNavTool_rootLevel_returnItemsOnlyForSite() throws Exception
...
+                .host(host).setProperty(FileAssetAPI.SHOW_ON_MENU, "true").nextPersisted();
+        final Template template = new TemplateDataGen().nextPersisted();
+        final Contentlet pageAssetOneHost = new HTMLPageDataGen(systemFolder, template)
+                .showOnMenu(true).host(host).nextPersisted();
+        ContentletDataGen.publish(pageAssetOneHost);
+        ContentletDataGen.publish(fileAssetOneHost);
+
+        new NavTool().getNav(host, systemFolder.getPath());
+        systemFolder = APILocator.getFolderAPI().findSystemFolder();
+        assertEquals(Host.SYSTEM_HOST, systemFolder.getHostId());
```
Comment (freddyucv) [inline dotCMS/src/integration-test/java/com/dotcms/rendering/velocity/viewtools/navigation/NavToolTest.java:481]: I think we should check also the return value of the getNav method

**Review by freddyucv — CHANGES_REQUESTED** [review]

**Review by nollymar — APPROVED** [review]

File: dotCMS/src/integration-test/java/com/dotcms/rendering/velocity/viewtools/navigation/NavToolTest.java (hunk 4 above)
Comment (dsilvam) [reply dotCMS/src/integration-test/java/com/dotcms/rendering/velocity/viewtools/navigation/NavToolTest.java:481]: It's covered in a different test

**Review by freddyucv — APPROVED** [review]

File: dotCMS/src/integration-test/java/com/dotcms/rendering/velocity/viewtools/navigation/NavToolTest.java (hunk 2 above)
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/rendering/velocity/viewtools/navigation/NavToolTest.java:211]: Codacy Issue found: [Local variable 'modifiedSystemFolder' could be declared final](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5378894)

File: dotCMS/src/main/java/com/dotcms/rendering/velocity/viewtools/navigation/NavTool.java [hunk 5]
```diff
@@ -77,13 +79,17 @@ protected NavResultHydrated getNav(Host host, String path, long languageId, User
             path = path.substring(0, path.lastIndexOf("/"));
         }
 
-        Folder folder = !path.equals("/") ? APILocator.getFolderAPI()
+        final Folder originalFolder = !path.equals("/") ? APILocator.getFolderAPI()
```
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotcms/rendering/velocity/viewtools/navigation/NavTool.java:82]: Codacy Issue found: [Position literals first in String comparisons](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5378894)

**Review by erickgonzalez — APPROVED** [review]

File: dotCMS/src/main/java/com/dotcms/rendering/velocity/viewtools/navigation/NavTool.java [hunk 6]
```diff
@@ -215,6 +221,22 @@ protected NavResultHydrated getNav(Host host, String path, long languageId, User
         }
     }
 
+    /**
+     * Makes a defensive copy of the given folder to avoid mutating cached version
+     * @param originalFolder folder to make the defensive copy from
+     * @return defensive copy
+     */
+    private Folder getDefensiveCopyOfFolder(Folder originalFolder) {
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/rendering/velocity/viewtools/navigation/NavTool.java:229]: Very good

**Review by jdotcms — APPROVED** [review]

**Pull request merged**